Add PHP support using php-cs-fixer

jr-cologne opened the pull request:
This PR adds PHP support by making use of the formatting tool `php-cs-fixer`.

Unfortunately, it does not work yet as `php-cs-fixer` automatically uses the config from `.php_cs`.
Because of that, the wanted file in `storage/code` is not formatted as the specified files from the config file have precedence over the given file in the execution command.
At least, this is what I think is causing that the specified file is not formatted.

To fix this, I have tried to "docker-ignore" `.php_cs`. However, this does not work either.
In general, it looks like `.dockerignore` has no effect on which files are copied into the container.

Does anyone have an idea how to solve this or what is actually going on?

jr-cologne committed:
Add php support using php-cs-fixer

asccc commented on .dockerignore:
```diff
@@ -7,3 +7,8 @@ LICENSE
 composer.json
 composer.lock
+.php_cs
+.php_cs.cache
```
php_cs.cache should be added to .gitignore instead

asccc commented on app/CodeFormatter/PhpCodeFormatter.php:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * @var array
+     */
+    protected static $supported_languages = [
```
i recently changed this static variable to a constant

```suggestion
    protected const LANGUAGES = [
```

asccc commented on app/CodeFormatter/PhpCodeFormatter.php:
```diff
@@ -0,0 +1,26 @@
+     */
+    protected function getShellCommand(string $file): string
+    {
+        return 'php-cs-fixer fix ' . $file;
```
i recently rewrote the formatter-logic to use pipes instead of files.

```suggestion
        return 'php-cs-fixer fix --rules @PSR2 -';
```

asccc commented on app/CodeFormatter/PhpCodeFormatter.php:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @return string
+     */
+    protected function getShellCommand(string $file): string
```
```suggestion
    protected function getShellCommand(): string
```

asccc commented on app/CodeFormatter/PhpCodeFormatter.php:
```diff
@@ -0,0 +1,26 @@
+    ];
+
+    /**
+     * @param string $file
```
remove this param-attribute to make phan happy 😆

asccc reviewed (changes requested):
the following suggestions are based on my recent pull-request (https://github.com/dev-community-de/code-formatter/pull/27) - which should be merged before this pr, otherwise things will break 😃

we can pass rules as an argument to make php-cs-fixer ignore our php_cs file

jr-cologne committed:
Apply changes to comply with PR #27

jr-cologne commented:
Thank you for your help! 👍

jr-cologne committed:
Merge branch 'master' into php-support

jr-cologne commented:
Oops, we cannot merge this yet.

I just encountered more issues with regards to the formatting of php code and the config files of `php-cs-fixer`.
However, there is a bug which is a much bigger deal:
In case formatting fails and an exception is thrown or the code formatter writes something to stdout/stderr, this output is printed / replaces parts of the actual post.
This means, instead of showing the formatted code, everyone in the forum sees the output (e.g. PHP exception) and the code posted in the forum might be lost.
This can potentially reveal sensitive information and is definitely a security vulnerability, not only an annoying and very uncommon bug.
I don't know yet where this is coming from or rather why this is happening exactly, but this most likely has something to do with PR #27.

Example:

Input:
```
Hallo
[CODE=php]<?php

class Test{
    private prop = '' ;

    public function test() {

        return      'test';
    }
}[/CODE]
test
```

```
Hallo
<br />
<b>Fatal error</b>:  Uncaught DevCommunityDE\CodeFormatter\Exceptions\Exception: code formatting failed! stderr: Loaded config default.
Using cache file &quot;.php_cs.cache&quot;.

Files that were not fixed due to errors reported during linting before fixing:
   1) stdin.php
 in /srv/app/app/CodeFormatter/CodeFormatter.php:82
Stack trace:
#0 /srv/app/app/CodeFormatterApp.php(74): DevCommunityDE\CodeFormatter\CodeFormatter\CodeFormatter-&gt;exec('&lt;?php\r\n\r\nclass ...')
#1 /srv/app/app/CodeFormatterApp.php(38): DevCommunityDE\CodeFormatter\CodeFormatterApp-&gt;formatToken(Object(DevCommunityDE\CodeFormatter\Parser\Token))
#2 /srv/app/bootstrap/app.php(12): DevCommunityDE\CodeFormatter\CodeFormatterApp-&gt;run()
#3 /srv/app/public/index.php(3): require_once('/srv/app/bootst...')
#4 {main}
  thrown in <b>/srv/app/app/CodeFormatter/CodeFormatter.php</b> on line <b>82</b><br />
```

jr-cologne commented:
Hmm, I'm still getting an error with regards to the php-cs-fixer config/cache files:

```
stderr: Loaded config default. Using cache file ".php_cs.cache".
```

Any idea why this is?

asccc commented on app/CodeFormatter/PhpCodeFormatter.php:
```diff
@@ -0,0 +1,21 @@
+     */
+    protected function getShellCommand(): string
+    {
+        return 'php-cs-fixer fix --rules @PSR2 -';
```
Lets try this flag:
```suggestion
        return 'php-cs-fixer fix --rules @PSR2 --quiet -';
```

And if this is not enough, we can redirect stderr to /dev/null

```suggestion
        return 'php-cs-fixer fix --rules @PSR2 --quiet - 2>/dev/null';
```

asccc reviewed (dismissed):
The formatter complains, because php-cs-fixer writes stuff to stderr,
I made some suggestions to fix this, but i have not tested them.

The fact that php-cs-fixer used its default config is good, because this means that it did not load our .php_cs file.

asccc commented:
Forget what i said. php-cs-fixer does not support stdin/stdout fixing (using pipes). 

It supports file-content from stdin, but does not write the result to stdout ... it only supports "diff" ...
We could apply the diff to the content but that would make things unnecessarily complicated.

Instead we could try the php formatter from prettier. 
The prettier php-formater is not as sophisticated as php-cs-fixer, but probably good enough™. And to be honest, in most cases, we just want to fix indentations in user provided scripts and not enforce PSR standards to it, because this could lead to situations where users do not recognize their own code anymore 😆

jr-cologne commented:
Yeah, ok. I kind of suspected that already.

For the future, it might be a good idea to support both ways of input/output - to a temp file as well as stdin/stdout using pipes.

In https://github.com/dev-community-de/code-formatter/pull/11, @wookiefriseur mentioned that Prettier plugins are way to slow which is why we decided not to use Prettier for formatting PHP.
Of course, we could still try it again.

asccc commented:
Since https://github.com/dev-community-de/code-formatter/pull/39 was merged (🎉) we could take look at [prettier-plugin-php](https://github.com/prettier/plugin-php) again.
Maybe the performance is reasonable now?

jr-cologne commented:
Yep, I am going to work on that pretty soon.
After merging this PR, I will finally release and deploy a new version to production.

jr-cologne committed:
Merge branch 'master' into php-support

jr-cologne committed:
Fix #40

This change ensures the original code-body is returned if code formatting fails.

jr-cologne committed:
Implement php support with prettier

jr-cologne commented:
I think we are good to go. 😃 

Prettier is significantly faster than before.
In manual testing, disabling `prettierd` raised the response time from around 100ms to 700ms.

asccc commented:
Nice, this sounds very promising!

jr-cologne commented:
Oops, merged it with the wrong title/description as we are not using `php-cs-fixer` anymore. 😄
